store/store.go

```diff
@@ -513,43 +505,35 @@ func (s *Store[H]) get(ctx context.Context, hash header.Hash) ([]byte, error) {
 	return data, nil
 }
 
-// try advance heighest header if we saw a higher continuous before.
-func (s *Store[H]) tryAdvanceHead(headers ...H) {
-	headPtr := s.writeHead.Load()
-	if headPtr == nil || len(headers) == 0 {
+// try advance heighest writeHead based on passed or already written headers.
+func (s *Store[H]) tryAdvanceHead(ctx context.Context, headers ...H) {
+	writeHead := s.writeHead.Load()
+	if writeHead == nil || len(headers) == 0 {
 		return
 	}
 
-	s.knownHeadersLk.Lock()
-	defer s.knownHeadersLk.Unlock()
+	currHeight := (*writeHead).Height()
 
-	for _, h := range headers {
-		s.knownHeaders[h.Height()] = h
+	// advance based on passed headers.
+	for i := 0; i < len(headers); i++ {
+		if headers[i].Height() != currHeight+1 {
+			break
+		}
+		s.writeHead.Store(&headers[i])
+		currHeight++
 	}
 
-	currHead := *headPtr
-	height := currHead.Height()
-	newHead := currHead
+	ctx, cancel := context.WithTimeout(ctx, time.Second)
+	defer cancel()
 
-	// try to move to the next height.
-	for len(s.knownHeaders) > 0 {
-		h, ok := s.knownHeaders[height+1]
-		if !ok {
+	// advance based on already written headers.
+	for {
+		newHead, err := s.GetByHeight(ctx, currHeight+1)
```

Comment:
Consider using [Query](https://github.com/ipfs/go-datastore/blob/fd1090a442b596704ef1512da4d8cb8bc7c98116/datastore.go#L94). It's a lower-level functionality that should presumably make this operation faster, although we would have to redo deserialization here.

Reply:
I am looking for a good example showing how to use Query in a similar scenario, but can't find the good one so far